org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/IDebugProvider.java

```diff
@@ -0,0 +1,15 @@
+/*******************************************************************************
+* Copyright (c) 2017 Microsoft Corporation and others.
+* All rights reserved. This program and the accompanying materials
+* are made available under the terms of the Eclipse Public License v1.0
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v10.html
+*
+* Contributors:
+*     Microsoft Corporation - initial API and implementation
+*******************************************************************************/
+
+package org.eclipse.jdt.ls.debug.adapter;
+
+public interface IDebugProvider extends ISourceLookUpProvider, IVirtualMachineManagerProvider {
```

Comment:
Recommend IProviderContext {
 ISourceLookUpProvider getSourceLookUpProvider();
 IVirtualMachineManagerProvider getVirualMachineManagerProvider();
}

In this way, we compose different provider implementations instead of forcing implementing all functions in the class which implements IDebugProvider.

Check:
https://www.safaribooksonline.com/library/view/effective-java-2nd/9780137150021/ch04.html
Item 16: Favor composition over inheritance